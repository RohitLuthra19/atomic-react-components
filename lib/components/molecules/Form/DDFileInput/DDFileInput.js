// @flow
import React, { useState } from 'react';
import type { Node } from 'react';
import { Field } from 'formik';
import styled from 'styled-components';

import Input from '../../../atoms/FileInput';
import styles from './DDFileInput.style';

type Props = {
  id: string,
  className?: string,
  type?: string,
  name: string,
  label: Node,
  labelProps?: Object,
  fieldProps?: Object,
  value?: string | number,
  placeholder?: string,
  onChange?: Function,
  onBlur?: Function,
};

const DDFieldInput = ({
  id,
  className,
  type,
  name,
  label,
  onChange,
  onBlur,
  labelProps,
  fieldProps,
  ...others
}: Props): Node => {
  const fileInput = React.createRef();
  const [highlight, setHighlight] = useState(false);
  let styleCls = className;

  function openFileDialog() {
    fileInput.current.click();
  }

  function fileListToArray(list) {
    const array = [];
    for (let i = 0; i < list.length; i += 1) {
      array.push(list.item(i));
    }
    return array;
  }

  function onDragOver(evt) {
    evt.preventDefault();
    setHighlight(true);
  }

  function onDragLeave() {
    setHighlight(false);
  }

  function onDrop(event) {
    event.preventDefault();
    const { files } = event.dataTransfer;
    const array = fileListToArray(files);
    setHighlight(false);
    return array;
  }

  function onFilesAdded(evt) {
    const { files } = evt.target;
    const array = fileListToArray(files);
    return array;
  }

  function handleKeyPress(evt) {
    // handle event
    if (evt.keyCode === 13) {
      openFileDialog();
    }
  }

  if (highlight) {
    styleCls = `${className} highLight`;
  }

  return (
    // TODO: Add support for label position
    <div
      className={styleCls}
      onClick={openFileDialog}
      onDragOver={onDragOver}
      onDragLeave={onDragLeave}
      onDrop={onDrop}
      onKeyDown={handleKeyPress}
      role="button"
      tabIndex="0"
    >
      <Field
        id={id}
        name={name}
        type={type}
        render={({ field, form }) => {
          const { handleChange, handleBlur } = form;
          return (
            <Input
              type={type}
              ref={fileInput}
              {...others}
              {...field}
              onChange={e => {
                handleChange(e);
                onFilesAdded(e);
                if (typeof onChange === 'function') onChange(e, form);
              }}
              onBlur={e => {
                handleBlur(e);
                if (typeof onBlur === 'function') onBlur(e, form);
              }}
            />
          );
        }}
        {...fieldProps}
      />
      Drop Files Here...
    </div>
  );
};

DDFieldInput.defaultProps = {
  placeholder: '',
  className: '',
  type: 'file',
  labelProps: {},
  value: '',
  fieldProps: {},
  onChange: null,
  onBlur: null,
};

export default styled(DDFieldInput)`
  ${styles};
`;

export { DDFieldInput };
