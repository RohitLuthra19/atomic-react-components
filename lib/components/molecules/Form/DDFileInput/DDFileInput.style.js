import { css } from 'styled-components';

// TODO: Fix the duplicate styling in input.style.js
const styles = css`
  margin-bottom: 0.75rem;
  border: 1px dashed grey;
  display: flex;
  justify-content: space-between;
  padding: 5px;
  input {
    // display: none;
  }
`;

export default styles;
