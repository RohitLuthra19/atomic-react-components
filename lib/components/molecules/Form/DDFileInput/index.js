// @flow

export { default } from './DDFileInput';
