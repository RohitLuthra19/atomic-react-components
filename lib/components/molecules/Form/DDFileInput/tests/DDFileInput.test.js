import React from 'react';

import { shallowWithTheme, renderWithTheme } from '../../../../../../__mock__/themeMock';
import { DDFieldInput } from '../DDFileInput';

describe('DDFieldInput Component', () => {
  let ddField = '';
  beforeEach(() => {
    ddField = shallowWithTheme(<DDFieldInput />);
  });

  test('should render correctly', () => {
    expect(ddField).toMatchSnapshot();
  });

  test('should render Input component with proper type for multiple file upload.', () => {
    const fileMultipleInputProps = {
      name: 'fileupload',
      type: 'file',
      multiple: true,
    };
    ddField = renderWithTheme(<DDFieldInput {...fileMultipleInputProps} />);
    expect(ddField.find('input[type=file]')).toHaveLength(1);
  });
});
