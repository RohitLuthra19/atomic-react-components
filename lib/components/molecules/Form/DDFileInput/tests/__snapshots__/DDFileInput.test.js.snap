// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DDFieldInput Component should render correctly 1`] = `
<div
  className=""
  onClick={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex="0"
>
  <Field
    name="fileupload"
    render={[Function]}
    type="file"
  />
  Drop Files Here...
</div>
`;
